features/telemetry/protocols/bmp/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/parser/BmpParser.java

```diff
@@ -187,25 +195,39 @@ public Handler accept(final InetSocketAddress remoteAddress,
 
                 LOG.trace("Got packet: {}", packet);
 
+                // Build the message from the received packet
+                final Transport.Message.Builder message = Transport.Message.newBuilder()
+                                                                           .setVersion(header.version);
+
+                packet.accept(new Serializer(message));
+
                 packet.accept(new Packet.Visitor.Adapter() {
                     @Override
-                    public void visit(InitiationPacket packet) {
+                    public void visit(final InitiationPacket packet) {
                         packet.information.first(InformationElement.Type.BGP_ID)
-                                .map(InetAddressUtils::addr)
-                                .ifPresent(_bgpId -> bgpId = _bgpId);
+                                          .map(InetAddressUtils::addr)
+                                          .ifPresent(_bgpId -> bgpId = _bgpId);
                     }
                 });
 
-                final Transport.Message.Builder message = Transport.Message.newBuilder()
-                                                                           .setVersion(header.version);
-
                 if (bgpId != null) {
-                    message.setBgpId(address(bgpId));
+                    message.setBgpId(BmpParser.address(bgpId));
                 }
 
-                packet.accept(new Serializer(message));
 
-                final CompletableFuture<AsyncDispatcher.DispatchStatus> dispatched = dispatcher.send(new TelemetryMessage(remoteAddress, ByteBuffer.wrap(message.build().toByteArray())));
+                // Enrich the message with resolved hostnames
+                CompletableFuture<Transport.Message.Builder> enriched = CompletableFuture.completedFuture(message);
+                if (BmpParser.this.dnsLookupsEnabled) {
+                    enriched = enriched.thenCompose(BmpParser.this.resolvePeer(packet))
```

Comment:
We need to find a way to put backpressure on the listener if we're not able to keep up. If there are already K enrichments in flight, then we should block until one of these finishes, otherwise we can OOM.

We handle this in `org.opennms.netmgt.telemetry.protocols.netflow.parser.ParserBase`, rather non-intutively. Maybe we can find a better way here? We already have resilience4j in the class-path - we could try and make use of the bulkhead: https://resilience4j.readme.io/docs/bulkhead